### PR TITLE
Test expired metadata from cache

### LINKED ISSUE #1681: ngtests: test loading of cached metadata
**Description of issue or feature request**:
Taken from @jku's [comment](https://github.com/theupdateframework/python-tuf/pull/1636#issuecomment-953669058) that summarizes it very well,  tests are needed which:

- check that expected metadata is read from cache the expected number of times
     - wrapping open() should work fine here (see tests in #1593)
-  check if cached metadata is used to verify new metadata downloaded from remote
      - trickier but I suppose we can rely on negative tests here: tests that are expected to fail on that verification

### PR DESCRIPTION
**Pull request by ivanayov**
This tests that an expired timestamp/snapshot/targets when loaded
from cache is not stored as final but is used to verify the new
timestamp

Addresses #1681

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [n/a] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by ivanayov** [comment]
I'm leaving this with the non-matching description on purpose, as the expected behaviour in
the issue description is to load expired snapshot/targets, which is not the case. Not loading them
is actually correct from what I understand from the spec - you shouldn't be able to use outdated timestamp for verification, thus you cannot verify snapshot and targets as well. Please, correct me if I misunderstand. 
So we could ether reword the requirement, of if there's a bug we can fix it.

File: tests/test_updater_with_simulator.py [hunk 1]
```diff
@@ -257,6 +259,50 @@ def test_not_loading_targets_twice(self, wrapped_open: MagicMock) -> None:
...
+        with self.assertRaises(ExpiredMetadataError):
+            updater.get_targetinfo(self.targets_dir)
+
+        # Test that an expired timestamp/snapshot/targets
+        # when loaded from cache is not stored as final
+        # but is used to verify the new timestamp
+        wrapped_open.assert_has_calls([
+            call(os.path.join(self.metadata_dir, "root.json"), "rb"),
+            call(os.path.join(self.metadata_dir, "timestamp.json"), "rb"),
+        ])
```
Comment (MVrachev) [inline tests/test_updater_with_simulator.py]: When reading the changes made in this pr I am comparing them to @sechkova comment here: https://github.com/theupdateframework/python-tuf/issues/1681#issuecomment-977851021.
It's not clear to me after you finish step 2 which steps do you cover with those asserts.
@ivanayov can you give more details on that?

Comment (ivanayov) [reply tests/test_updater_with_simulator.py]: @MVrachev This PR is for step 4 only. The others are covered in #1703 .

Comment (ivanayov) [reply tests/test_updater_with_simulator.py]: Step 5 is already covered in the tests, so we discussed with @jku that it's not necessary to repeat it here.

Comment (MVrachev) [reply tests/test_updater_with_simulator.py]: I will look at it today a little later.

Comment (MVrachev) [reply tests/test_updater_with_simulator.py]: Please keep in mind that by using `Fixes` in your pr description you are going to close the issue you are referencing with this pr.
That's okay if you handle all of the steps here, but as it seems you don't test for all of the steps here, but you also rely on #1703 in order to do that.
Now the first pr that merges between this one and #1703 will close the issue even though none of the two prs fixes the issue entirely by itself.

File: tests/test_updater_with_simulator.py [hunk 2]
```diff
@@ -257,6 +259,50 @@ def test_not_loading_targets_twice(self, wrapped_open: MagicMock) -> None:
...
+            "https://example.com/targets/",
+            self.sim,
+        )
+
+        with self.assertRaises(ExpiredMetadataError):
+            updater.get_targetinfo(self.targets_dir)
+
+        # Test that an expired timestamp/snapshot/targets
+        # when loaded from cache is not stored as final
+        # but is used to verify the new timestamp
```
Comment (MVrachev) [inline tests/test_updater_with_simulator.py]: Here you mentioned that you are checking that an expired `timestamp/snapshot/targets, when loaded from cache, is not stored as final but is used to verify the new timestamp`, but you are only checking that there are
calls for `root.json` and `timestamp.json`.
I thought from your comment and step 3 that you will have to check for `snapshot` and `targets` as well.

Comment (ivanayov) [reply tests/test_updater_with_simulator.py]: @MVrachev I've described this in https://github.com/theupdateframework/python-tuf/pull/1707#issuecomment-985324978 . Any feedback on that direction is welcome.

File: tests/test_updater_with_simulator.py (hunk 1 above)
Comment (ivanayov) [reply tests/test_updater_with_simulator.py]: Thanks, I've edited the description.

File: tests/test_updater_with_simulator.py (hunk 2 above)
Comment (MVrachev) [reply tests/test_updater_with_simulator.py]: I am sorry that I didn't read your comment above. I have read this pr description before the comment was posted and didn't check if there are new comments...

File: tests/test_updater_with_simulator.py [hunk 3]
```diff
@@ -257,6 +259,50 @@ def test_not_loading_targets_twice(self, wrapped_open: MagicMock) -> None:
...
+        updater = Updater(
+            self.metadata_dir,
+            "https://example.com/metadata/",
+            self.targets_dir,
+            "https://example.com/targets/",
+            self.sim,
+        )
+
+        with self.assertRaises(ExpiredMetadataError):
+            updater.get_targetinfo(self.targets_dir)
```
Comment (MVrachev) [inline tests/test_updater_with_simulator.py]: Why did you call `updater.get_targetinfo(self.targets_dir)`?
Why didn't you use directly `updater.refresh()`?
Also, the argument to `updater.get_targetinfo()` should be a path to a specific target, not the targets directory.

Comment (MVrachev) [reply tests/test_updater_with_simulator.py]: I asked Teodora about this here https://github.com/theupdateframework/python-tuf/issues/1681#issuecomment-987330543,

**Comment by MVrachev** [comment]
I have some ideas about this pr and will have a conversation with Ivana soon.

**Comment by jku** [comment]
> the expected behaviour in
the issue description is to load expired snapshot/targets, which is not the case. Not loading them
is actually correct from what I understand from the spec 

This may have been clarified already but I'l say it here:

We do load expired metadata with one big caveat: the expired metadata can only be used to validate a newer version of the _same type of metadata_. In practice:
* local, possibly expired, metadata is loaded and used to validate newer downloaded metadata. "validate" here means signature verification for root and various version checks for timestamp, snapshot and targets.
* even the downloaded versions may be expired (in practice only root is, because we download all root versions up to the current, hopefully non-expired version)
* the python-tuf implementation  will fail the update once the process moves to the next type of metadata and the previous metadata is still expired (e.g. if final root version is expired, refresh() will fail when it tries to validate timestamp)

**Comment by ivanayov** [comment]
`tox` is giving really strange proposals here:
```
-        past_datetime = datetime.utcnow().replace(microsecond=0) - timedelta(days=5)
+        past_datetime = datetime.utcnow().replace(microsecond=0) - timedelta(
+            days=5
+        )
```
Do we really expect that or it's misconfiguration?

**Comment by jku** [comment]
the core issue here is that _verifying that local files are opened does not mean they are actually used to verify new downloaded metadata_ (this is made clear even in the test itself: it opens `targets.json` but we know that's not used for anything). This does not mean the test is worthless but at least some descriptions in the PR and commit are not describing it correctly. I would like to see a clear explanation of what the test does so that 
 A) we can be sure we don't already have a test that does the same checks
 B) future readers won't be deceived by inaccurate descriptions 
I get that this may be the hardest part of this PR: let's talk if you don't have a good idea already

Just to be clear: I think this is a useful test that probably does not exist yet... but test_updater_top_level_update.py has some clearly related tests like `test_new_timestamp_expired()`: maybe it should have a companion `test_local_timestamp_expired()`? I don't have strong opinions here but maybe @sechkova has ideas about where to put tests like this -- and whether other tests exist already?

Also (and this is a small detail in comparison), I've been trying to advocate not meddling with the client metadata cache manually: in this case I think it would be better to not do that and instead mock current time (pretend that time now > expiry). Search for mock_time in tests for examples of this.

File: tests/test_updater_with_simulator.py [hunk 4]
```diff
@@ -254,6 +255,32 @@ def test_not_loading_targets_twice(self, wrapped_open: MagicMock) -> None:
...
+
+        # Make a successful update of valid metadata which stores it in cache
+        self._run_refresh()
+
+        # Simulate expired local metadata by mocking system time one second ahead
+        mock_time = Mock()
+        mock_time.return_value = (
+            int(self.sim.timestamp.expires.strftime("%Y%m%d%H%M%S")) + 1
+        )
+        with patch("time.time", mock_time):
```
Comment (sechkova) [inline tests/test_updater_with_simulator.py]: I am not sure I understand the exact intentions but the tests pass because patch("time.time", mock_time) does not work as you expected and there isn't any expired metadata. The tests would work the exact same way if you delete the lines doing the "mocking". 
Updater (TrustedMetadataSet) uses datetime.utcnow to calculate the reference time.

**Comment by sechkova** [comment]
> Just to be clear: I think this is a useful test that probably does not exist yet... but test_updater_top_level_update.py has some clearly related tests like `test_new_timestamp_expired()`: maybe it should have a companion `test_local_timestamp_expired()`? I don't have strong opinions here but maybe @sechkova has ideas about where to put tests like this -- and whether other tests exist already?

I admit I find it hard to follow the intentions here, I am sorry if I've put a misleading issue description.
If I understand correctly, `test_trusted_root_expired` from test_updater_top_level_update.py is doing something very similar and can be used as an example.

**Comment by ivanayov** [comment]
@sechkova the idea was to simulate an expired metadata in cache without manually modifying the cache files. I might have misunderstood this mock - the intention is to simulate change in system time, so that what's stored in cache is considered expired, without need of directly touching the local metadata.

**Comment by jku** [comment]
> I admit I find it hard to follow the intentions here, I am sorry if I've put a misleading issue description. If I understand correctly, `test_trusted_root_expired` from test_updater_top_level_update.py is doing something very similar and can be used as an example.

It is pretty similar, except that I think we don't actually persist downloaded non-root metadata if it's expired, even though we probably could (rado pointed this out to me -- I had not noticed). This difference isn't that important for actual working repos but for testing it means we need to mock system time (or utcnow() as you point out) to "make the local metadata expired".

**Comment by sechkova** [comment]
> > I admit I find it hard to follow the intentions here, I am sorry if I've put a misleading issue description. If I understand correctly, `test_trusted_root_expired` from test_updater_top_level_update.py is doing something very similar and can be used as an example.
> 
> It is pretty similar, except that I think we don't actually persist downloaded non-root metadata if it's expired, even though we probably could (rado pointed this out to me -- I had not noticed). This difference isn't that important for actual working repos but for testing it means we need to mock system time (or utcnow() as you point out) to "make the local metadata expired".

Thanks, now I think I got all pieces together. So I agree, `test_updater_top_level_update.py` could host the additional tests for loading expired snapshot and timestamp and since expired metadata and timestamp are not persisted, mocking seems to be the way. 
I think expired local "targets" should be treated the same way as an invalid (unsigned) local metadata and may be part of another set of tests?

**Comment by jku** [comment]
I'm marking this a draft since the issue pointed out by Teodora (mocking the wrong thing) is still there: I might have other comments but I've been postponing review until the basics are there. I'm available for a chat if anything is unclear.

**Comment by ivanayov** [comment]
@jku I changed to mocking `datetime` as Teodora advised, so it's now available for review.

**Review by jku — CHANGES_REQUESTED** [review]
marking request changes so it's visible on PR list: we talked about this over zoom, and found an issue with the timeline of expiries and the mocked refresh: ivana promised to fix

**Comment by ivanayov** [comment]
I've pushed an update.

File: tests/test_updater_top_level_update.py [hunk 5]
```diff
@@ -306,6 +308,69 @@ def test_new_timestamp_unsigned(self) -> None:
 
         self._assert_files_exist([Root.type])
 
+    @patch.object(datetime, "datetime", wraps=datetime.datetime)
+    def test_expired_timestamp_version_rollback(self, mock_time: Mock) -> None:
+
+        now = datetime.datetime.utcnow()
+        self.sim.timestamp.expires = now.replace(
+            microsecond=0
+        ) + datetime.timedelta(days=7)
```
Comment (jku) [inline tests/test_updater_top_level_update.py]: I think I suggested this (since it's what repositorysimulator does) but in fact the `expires` setter nowadays handles the microseconds, so I think this works for all cases where the setter is used:

```suggestion
        self.sim.timestamp.expires = now + datetime.timedelta(days=7)
```

File: tests/test_updater_top_level_update.py [hunk 6]
```diff
@@ -306,6 +308,69 @@ def test_new_timestamp_unsigned(self) -> None:
...
+        self._run_refresh()
+
+        self.sim.timestamp.version = 1
+
+        self.sim.timestamp.expires = now.replace(
+            microsecond=0
+        ) + datetime.timedelta(days=21)
+
+        mock_time.utcnow.return_value = (
+            datetime.datetime.utcnow() + datetime.timedelta(days=18)
```
Comment (jku) [inline tests/test_updater_top_level_update.py:337]:
```suggestion
            now + datetime.timedelta(days=18)
```
maybe easier to make the connection to the expiry dates if we use the same construct.

This applies to the other tests as well

**Review by jku — CHANGES_REQUESTED** [review]
Nice, all test code seems correct and clear.

* Left some minor tweak suggestions in code
* These tests should have a docstring to explain what they are testing: they are some of the trickiest tests we have. Something like "Local timestamp must be used in rollback checks even if it is expired" -- maybe the same text that's already before the actual assertRaises

Additionally _maybe_ the dosctring should also include a timeline, example of one test:
```
  * day 0: first updater refresh
  * day 0 (a bit later): repository publishes timestamp v2
  * day 7: timestamp v1 expiry
  * day 18: second updater refresh: assert that rollback check uses expired v1 here
  * day 21: timestamp v2 expiry

```
not sure about that last thing though: it's a lot of lines... but may be helpful to understand what's going on? I'll leave decision to you

**Comment by ivanayov** [comment]
Thank you @jku , I added docstrings to all the test methods for clarification.

**Review by jku — APPROVED** [review]
my comment about reusing the `datetime.datetime.utcnow()` result  is marked resolved but all functions still call utcnow() twice... 

That is a trivial complaint so let's go with this version, I think it's a good set of very tricky tests. Thanks.

**Comment by ivanayov** [comment]
@jku ah, I resolved by inertia with the microseconds updates... will note to address in following changes. Thanks
